### PR TITLE
[GFX-592] Add ANGLE to Filament

### PR DESCRIPTION
**Pull request by marcisolti**
## Jira ticket URL (Why?)
[GFX-592](https://shapr3d.atlassian.net/browse/GFX-592)

## Short description (What? How?) 📖
You can now choose an ANGLE backend in Filament:
* A new preprocessor directive, `FILAMENT_USE_ANGLE` was introduced.
* ANGLE backend uses `PlatformEGL`, which was a bit modified to take an external `EGLDisplay` in the constructor.
  * We did not see much added value in implementing our own `PlatformEGL_ANGLE` as a descendant of `PlatformEGL`.
  * The only major difference would be, that we would not query the `EGL_RECORDABLE_ANDROID` config.
  * We did not have any issue so for leaving it in, so we do not deviate from upstream in this case.

Various small fixes for Win/OpenGL driver
* `OpenGLDriver::createRenderTargetR` did not call `glDrawBuffers` for depth-only passes.
* Commented out usage of `glTextParameteri` in `updateTextureLodRange` and `setMinMaxLevels`. [This card](https://shapr3d.atlassian.net/browse/GFX-1583) tricks this.
* Do not allow zero size render target in `getAttachments`
* C-style cast in `setExternalVertexBuffer` `setExternalIndexBuffer` from `intptr_t` to `GLeglClientBufferEXT` (from #46)

Usage:
* First, create an `EGLDeviceEXT` from an `ID3D11Device` with `eglCreateDeviceANGLE`.
* Create an `EGLDisplay` from the `EGLDeviceEXT` using `eglGetPlatformDisplayEXT`.
* Create a `PlatformEGL` with the `EGLDisplay`.
* Pass the `PlatformEGL*` to the `Engine` constructor.   

To run `shadowtest`, use branch `origin/GFX-592_add_angle_run_shadowtest`

### How did you test it? 🤔

Manual 💁‍♂️

Automated 💻

### CODE REVIEW

File: filament/backend/src/opengl/PlatformEGL.cpp [hunk 1]
```diff
@@ -75,11 +75,15 @@ static void clearGlError() noexcept {
 
 // ---------------------------------------------------------------------------------------------
 
-PlatformEGL::PlatformEGL() noexcept = default;
+PlatformEGL::PlatformEGL(EGLDisplay display = EGL_NO_DISPLAY) noexcept {
```
Comment (palver123) [inline filament/backend/src/opengl/PlatformEGL.cpp:78]: Declare the default parameter value in the header pls

File: filament/backend/src/Platform.cpp [hunk 2]
```diff
@@ -138,6 +143,8 @@ DefaultPlatform* DefaultPlatform::create(Backend* backend, void* nativeDevice) n
     #if defined(FILAMENT_SUPPORTS_OPENGL)
         #if defined(FILAMENT_USE_EXTERNAL_GLES3) || defined(FILAMENT_USE_SWIFTSHADER)
             return nullptr;
+        #elif defined(FILAMENT_USE_ANGLE)
+            return new PlatformEGL(nativeDevice);
```
Comment (palver123) [inline filament/backend/src/Platform.cpp:147]: Please rename the parameter. `nativeDisplay`? And don't forget to update the comment in Platform.h (we forgot to add a description last time)

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 3]
```diff
@@ -1079,18 +1079,16 @@ void OpenGLDriver::createRenderTargetR(Handle<HwRenderTarget> rth,
     rt->gl.samples = samples;
     rt->targets = targets;
 
+    GLenum bufs[MRT::MAX_SUPPORTED_RENDER_TARGET_COUNT] = { GL_NONE };
+    const size_t maxDrawBuffers = getMaxDrawBuffers();
```
Comment (palver123) [inline filament/backend/src/opengl/OpenGLDriver.cpp:1083]: Please add a comment about this in the PR description. It was a bug not to call glDrawBuffers for depth-only passes (used in post processing I think?)

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 4]
```diff
@@ -1640,7 +1639,7 @@ void OpenGLDriver::setExternalIndexBuffer(Handle<HwIndexBuffer> ibh, intptr_t ex
 
     gl.bindVertexArray(nullptr);
     gl.bindBuffer(GL_ELEMENT_ARRAY_BUFFER, ib->gl.id);
-    glBufferStorageExternalEXT(GL_ELEMENT_ARRAY_BUFFER, 0, ib->count * ib->elementSize, externalBuffer, 0);
+    glBufferStorageExternalEXT(GL_ELEMENT_ARRAY_BUFFER, 0, ib->count * ib->elementSize, (GLeglClientBufferEXT)externalBuffer, 0);
```
Comment (palver123) [inline filament/backend/src/opengl/OpenGLDriver.cpp:1642]: A quick note about this in the PR desc: leftover from #46

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 5]
```diff
@@ -1843,11 +1842,11 @@ void OpenGLDriver::setMinMaxLevels(Handle<HwTexture> th, uint32_t minLevel, uint
     // Must fit within int8_t.
     assert_invariant(minLevel <= 0x7f && maxLevel <= 0x7f);
 
-    t->gl.baseLevel = minLevel;
+    /*t->gl.baseLevel = minLevel;
     glTexParameteri(t->gl.target, GL_TEXTURE_BASE_LEVEL, t->gl.baseLevel);
 
     t->gl.maxLevel = maxLevel; // NOTE: according to the GL spec, the default value of this 1000
-    glTexParameteri(t->gl.target, GL_TEXTURE_MAX_LEVEL, t->gl.maxLevel);
+    glTexParameteri(t->gl.target, GL_TEXTURE_MAX_LEVEL, t->gl.maxLevel);*/
```
Comment (palver123) [inline filament/backend/src/opengl/OpenGLDriver.cpp:1849]: Add a note to the PR desc: Fixing a perf. issue (CPU copies inside ANGLE). [GFX-1583](https://shapr3d.atlassian.net/browse/GFX-1583) tracks this.

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 6]
```diff
@@ -2455,10 +2454,10 @@ GLsizei OpenGLDriver::getAttachments(AttachmentArray& attachments,
         assert_invariant(!defaultFramebuffer);
         attachments[attachmentCount++] = GL_COLOR_ATTACHMENT7;
     }
-    if (any(buffers & TargetBufferFlags::DEPTH)) {
+    if (any(buffers & TargetBufferFlags::DEPTH) && rt->width != 0 && rt->height != 0) {
```
Comment (palver123) [inline filament/backend/src/opengl/OpenGLDriver.cpp:2457]: Add a note to the PR desc. about this: 👴 I don't remember, ask @veghbalazs again please

File: filament/backend/include/backend/PixelBufferDescriptor.h [hunk 7]
```diff
@@ -247,7 +247,7 @@ class UTILS_PUBLIC PixelBufferDescriptor : public BufferDescriptor {
         }
 
         size_t bpr = bpp * stride;
-        size_t bprAligned = (bpr + (alignment - 1)) & -alignment;
+        size_t bprAligned = (bpr + (alignment - 1)) & (~alignment + 1);
```
Comment (palver123) [inline filament/backend/include/backend/PixelBufferDescriptor.h:250]: Is this necessary now? Do the Filament libs build without it in VS2022? If not necessary, I would postpone this (along with other MSVC warning hunt) in [GFX-1590](https://shapr3d.atlassian.net/browse/GFX-1590)

**Review by palver123 — COMMENTED** [review]
Note: The native device parameter of Engine ctor can be reverted now. But I think partially reverting the changes of #1 should be a separate PR. For example `SwapChain::setSwapInterval()` can be reverted too.

For now I track this task in [GFX-1590](https://shapr3d.atlassian.net/browse/GFX-1590). I suggested not doing these in this PR

File: filament/backend/src/opengl/PlatformEGL.cpp (hunk 1 above)
Comment (marcisolti) [reply filament/backend/src/opengl/PlatformEGL.cpp]: 780f261

File: filament/backend/src/Platform.cpp (hunk 2 above)
Comment (marcisolti) [reply filament/backend/src/Platform.cpp]: e9c0f52

File: filament/backend/src/opengl/OpenGLDriver.cpp (hunk 3 above)
Comment (marcisolti) [reply filament/backend/src/opengl/OpenGLDriver.cpp:1083]: Added comment in the description.

File: filament/backend/src/opengl/OpenGLDriver.cpp (hunk 5 above)
Comment (marcisolti) [reply filament/backend/src/opengl/OpenGLDriver.cpp:1849]: Added comment in the description.

File: filament/backend/src/opengl/OpenGLDriver.cpp (hunk 6 above)
Comment (marcisolti) [reply filament/backend/src/opengl/OpenGLDriver.cpp:2457]: Added comment in the description.

File: filament/backend/include/backend/PixelBufferDescriptor.h (hunk 7 above)
Comment (marcisolti) [reply filament/backend/include/backend/PixelBufferDescriptor.h]: Reverted in a17abd4

File: filament/backend/src/opengl/PlatformEGL.cpp [hunk 8]
```diff
@@ -75,11 +75,15 @@ static void clearGlError() noexcept {
 
 // ---------------------------------------------------------------------------------------------
 
-PlatformEGL::PlatformEGL() noexcept = default;
+PlatformEGL::PlatformEGL(EGLDisplay display) noexcept {
+    mEGLDisplay = display;
```
Comment (palver123) [inline filament/backend/src/opengl/PlatformEGL.cpp:79]: nit: this could go into initializer list

**Review by palver123 — APPROVED** [review]

File: filament/backend/src/opengl/PlatformEGL.cpp (hunk 8 above)
Comment (marcisolti) [reply filament/backend/src/opengl/PlatformEGL.cpp]: 1b30b28

File: filament/backend/src/opengl/OpenGLContext.h [hunk 9]
```diff
@@ -155,6 +155,7 @@ class OpenGLContext {
         bool WEBGL_compressed_texture_etc = false;
         bool WEBGL_compressed_texture_s3tc = false;
         bool WEBGL_compressed_texture_s3tc_srgb = false;
+        bool EXT_external_buffer = false;
```
Comment (veghbalazs) [inline filament/backend/src/opengl/OpenGLContext.h:158]: What's this, a leftover from https://github.com/shapr3d/filament/pull/1? External buffers aren't that relevant at this point, the ticket about those is [GFX-594](https://shapr3d.atlassian.net/browse/GFX-594).

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 10]
```diff
@@ -2455,10 +2454,10 @@ GLsizei OpenGLDriver::getAttachments(AttachmentArray& attachments,
...
         attachments[attachmentCount++] = defaultFramebuffer ? GL_DEPTH : GL_DEPTH_ATTACHMENT;
     }
-    if (any(buffers & TargetBufferFlags::STENCIL)) {
+    if (any(buffers & TargetBufferFlags::STENCIL) && rt->width != 0 && rt->height != 0) {
```
Comment (veghbalazs) [inline filament/backend/src/opengl/OpenGLDriver.cpp:2460]: Did you verify that we still need this extra check? It was more of a shortcut solution, and we should understand if 1) it still causes an assert in ANGLE and 2) how framebuffers or attachments of zero dimensions may be born.

File: filament/backend/src/opengl/OpenGLDriver.cpp [hunk 11]
```diff
@@ -3176,14 +3175,14 @@ void OpenGLDriver::updateTextureLodRange(GLTexture* texture, int8_t targetLevel)
         if (targetLevel < texture->gl.baseLevel || targetLevel > texture->gl.maxLevel) {
             bindTexture(OpenGLContext::MAX_TEXTURE_UNIT_COUNT - 1, texture);
             gl.activeTexture(OpenGLContext::MAX_TEXTURE_UNIT_COUNT - 1);
-            if (targetLevel < texture->gl.baseLevel) {
+            /*if (targetLevel < texture->gl.baseLevel) {
```
Comment (veghbalazs) [inline filament/backend/src/opengl/OpenGLDriver.cpp:3178]: a ticket number would be nice in comments for these types of "fixes", (same in `setMinMaxLevels()`)

File: third_party/angle/tnt/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+cmake_minimum_required(VERSION 3.19)
+project(angle)
+
+add_library(angle INTERFACE)
+target_include_directories(angle INTERFACE ${ANGLE_INCLUDE_DIR})
```
Comment (veghbalazs) [inline third_party/angle/tnt/CMakeLists.txt:5]: I would leave some commented code here, which helps anyone wanting to try out sample targets using ANGLE, (i.e. linking dependecies). For example:
```cmake
# Uncomment the following lines and specify your folder containing ANGLE libraries to enable linking sample projects using ANGLE:
# set(ANGLE_LIBS_DIR "")
# target_link_directories(angle INTERFACE ${ANGLE_LIBS_DIR})
# target_link_libraries(angle INTERFACE "libEGL.dll.lib" "libGLESv2.dll.lib")
```

File: filament/backend/include/backend/Platform.h [hunk 13]
```diff
@@ -76,15 +76,19 @@ class UTILS_PUBLIC DefaultPlatform : public Platform {
...
+     * @param nativeDisplay Specify a display to connect to. Currently only meaningful for the EGL
+     *                      platform, on which you should pass in an EGLDisplay if you want to use
+     *                      one different from the default.
      *
      * @return A pointer to the Plaform object.
      *
      * @see destroy
      */
-    static DefaultPlatform* create(backend::Backend* backendHint, void* nativeDevice = nullptr) noexcept;
+    static DefaultPlatform* create(backend::Backend* backendHint, void* nativeDisplay = nullptr) noexcept;
```
Comment (veghbalazs) [inline filament/backend/include/backend/Platform.h:91]: Is the class `PlatformEGL` public? (It seems so based on your PR description). If yes, the `nativeDevice`/`nativeDisplay` parameter can be removed, (it was an addition by us, which would go unused).

**Pull request merged**